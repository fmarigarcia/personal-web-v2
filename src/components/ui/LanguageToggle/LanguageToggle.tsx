import React from 'react';
import { useTranslation } from 'react-i18next';
import clsx from 'clsx';

interface LanguageToggleProps {
    className?: string;
}

export const LanguageToggle: React.FC<LanguageToggleProps> = ({
    className = '',
}) => {
    const { i18n } = useTranslation();

    const handleLanguageChange = (lang: 'en' | 'es') => {
        i18n.changeLanguage(lang);
    };

    const getButtonClasses = (lang: 'en' | 'es') => {
        return clsx('text-sm font-medium px-2 py-1 rounded transition-colors', {
            'bg-blue-600 text-white': i18n.language === lang,
            'text-gray-700 hover:text-blue-600': i18n.language !== lang,
        });
    };

    return (
        <div className={clsx('flex items-center space-x-2', className)}>
            <button
                onClick={() => handleLanguageChange('en')}
                className={getButtonClasses('en')}
                type="button"
                aria-label="Switch to English"
            >
                EN
            </button>
            <button
                onClick={() => handleLanguageChange('es')}
                className={getButtonClasses('es')}
                type="button"
                aria-label="Switch to Spanish"
            >
                ES
            </button>
        </div>
    );
};

export default LanguageToggle;
