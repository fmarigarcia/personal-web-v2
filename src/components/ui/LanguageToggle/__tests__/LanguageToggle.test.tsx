import { render, screen, fireEvent } from '@testing-library/react';
import { LanguageToggle } from '../LanguageToggle';

// Mock i18next
const mockChangeLanguage = jest.fn();
jest.mock('react-i18next', () => ({
    useTranslation: () => ({
        i18n: {
            language: 'en',
            changeLanguage: mockChangeLanguage,
        },
    }),
}));

describe('LanguageToggle', () => {
    beforeEach(() => {
        mockChangeLanguage.mockClear();
    });

    it('should render language toggle buttons with flags', () => {
        render(<LanguageToggle />);

        expect(
            screen.getByRole('button', { name: /switch to english/i })
        ).toBeInTheDocument();
        expect(
            screen.getByRole('button', { name: /switch to spanish/i })
        ).toBeInTheDocument();

        // Check that flags are displayed
        expect(screen.getByText('ğŸ‡ºğŸ‡¸')).toBeInTheDocument();
        expect(screen.getByText('ğŸ‡ªğŸ‡¸')).toBeInTheDocument();
    });

    it('should render custom languages when provided', () => {
        const customLanguages = [
            { code: 'fr', flag: 'ğŸ‡«ğŸ‡·', label: 'French' },
            { code: 'de', flag: 'ğŸ‡©ğŸ‡ª', label: 'German' },
        ];

        render(<LanguageToggle languages={customLanguages} />);

        expect(
            screen.getByRole('button', { name: /switch to french/i })
        ).toBeInTheDocument();
        expect(
            screen.getByRole('button', { name: /switch to german/i })
        ).toBeInTheDocument();
        expect(screen.getByText('ğŸ‡«ğŸ‡·')).toBeInTheDocument();
        expect(screen.getByText('ğŸ‡©ğŸ‡ª')).toBeInTheDocument();
    });

    it('should highlight active language', () => {
        render(<LanguageToggle />);

        const enButton = screen.getByRole('button', {
            name: /switch to english/i,
        });
        const esButton = screen.getByRole('button', {
            name: /switch to spanish/i,
        });

        // English should be active by default
        expect(enButton).toHaveClass('bg-blue-600', 'text-white');
        expect(esButton).toHaveClass('text-gray-700');
    });

    it('should handle language change to Spanish', () => {
        render(<LanguageToggle />);

        const esButton = screen.getByRole('button', {
            name: /switch to spanish/i,
        });
        fireEvent.click(esButton);

        expect(mockChangeLanguage).toHaveBeenCalledWith('es');
    });

    it('should handle language change to English', () => {
        render(<LanguageToggle />);

        const enButton = screen.getByRole('button', {
            name: /switch to english/i,
        });
        fireEvent.click(enButton);

        expect(mockChangeLanguage).toHaveBeenCalledWith('en');
    });

    it('should handle custom language changes', () => {
        const customLanguages = [
            { code: 'fr', flag: 'ğŸ‡«ğŸ‡·', label: 'French' },
            { code: 'de', flag: 'ğŸ‡©ğŸ‡ª', label: 'German' },
        ];

        render(<LanguageToggle languages={customLanguages} />);

        const frButton = screen.getByRole('button', {
            name: /switch to french/i,
        });
        fireEvent.click(frButton);

        expect(mockChangeLanguage).toHaveBeenCalledWith('fr');
    });

    it('should apply custom className', () => {
        const customClass = 'custom-language-toggle';
        const { container } = render(
            <LanguageToggle className={customClass} />
        );

        const toggleContainer = container.querySelector('div');
        expect(toggleContainer).toHaveClass(customClass);
    });

    it('should have proper accessibility attributes', () => {
        render(<LanguageToggle />);

        const buttons = screen.getAllByRole('button');
        buttons.forEach((button) => {
            expect(button).toHaveAttribute('type', 'button');
            expect(button).toHaveAttribute('aria-label');
        });
    });

    it('should have hover styles for inactive buttons', () => {
        render(<LanguageToggle />);

        const esButton = screen.getByRole('button', {
            name: /switch to spanish/i,
        });
        expect(esButton).toHaveClass(
            'hover:text-blue-600',
            'transition-colors'
        );
    });

    it('should handle single language array', () => {
        const singleLanguage = [{ code: 'en', flag: 'ğŸ‡ºğŸ‡¸', label: 'English' }];

        render(<LanguageToggle languages={singleLanguage} />);

        expect(
            screen.getByRole('button', { name: /switch to english/i })
        ).toBeInTheDocument();
        expect(screen.getByText('ğŸ‡ºğŸ‡¸')).toBeInTheDocument();
        expect(screen.getAllByRole('button')).toHaveLength(1);
    });

    it('should match snapshot with default languages', () => {
        const { container } = render(<LanguageToggle />);
        expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot with custom languages', () => {
        const customLanguages = [
            { code: 'fr', flag: 'ğŸ‡«ğŸ‡·', label: 'French' },
            { code: 'de', flag: 'ğŸ‡©ğŸ‡ª', label: 'German' },
        ];

        const { container } = render(
            <LanguageToggle languages={customLanguages} />
        );
        expect(container.firstChild).toMatchSnapshot();
    });
});
