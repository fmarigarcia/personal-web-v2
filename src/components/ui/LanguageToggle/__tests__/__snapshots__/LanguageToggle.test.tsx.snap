// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageToggle should match snapshot 1`] = `
<div
  class="flex items-center space-x-2"
>
  <button
    aria-label="Switch to English"
    class="text-sm font-medium px-2 py-1 rounded transition-colors bg-blue-600 text-white"
    type="button"
  >
    EN
  </button>
  <button
    aria-label="Switch to Spanish"
    class="text-sm font-medium px-2 py-1 rounded transition-colors text-gray-700 hover:text-blue-600"
    type="button"
  >
    ES
  </button>
</div>
`;
